receiver/mongodbreceiver/client.go

```diff
@@ -0,0 +1,222 @@
+// Copyright  The OpenTelemetry Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package mongodbreceiver // import "github.com/open-telemetry/opentelemetry-collector-contrib/receiver/mongodbreceiver"
+
+import (
+	"context"
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"strings"
+	"time"
+
+	"go.mongodb.org/mongo-driver/bson"
+	"go.mongodb.org/mongo-driver/mongo"
+	"go.mongodb.org/mongo-driver/mongo/options"
+	"go.mongodb.org/mongo-driver/mongo/readpref"
+	"go.uber.org/zap"
+)
+
+// Client is an interface that exposes functionality towards a mongo environment
+type Client interface {
+	ListDatabaseNames(ctx context.Context, filters interface{}, opts ...*options.ListDatabasesOptions) ([]string, error)
+	Disconnect(context.Context) error
+	Connect(context.Context) error
+	TestConnection(context.Context) (*mongoTestConnectionResponse, error)
+	ServerStatus(ctx context.Context, DBName string) (bson.M, error)
+	DBStats(ctx context.Context, DBName string) (bson.M, error)
+}
+
+// driver is the underlying mongo.Client that Client invokes to actually do actions
+// against the mongo instance
+type driver interface {
+	Database(dbName string, opts ...*options.DatabaseOptions) *mongo.Database
+	ListDatabaseNames(ctx context.Context, filters interface{}, opts ...*options.ListDatabasesOptions) ([]string, error)
+	Connect(context.Context) error
+	Disconnect(context.Context) error
+	Ping(ctx context.Context, rp *readpref.ReadPref) error
+}
+
+// mongodbClient is a mongodb metric scraper client
+type mongodbClient struct {
+	// underlying mongo client driver
+	driver         driver
+	username       string
+	password       string
+	hosts          []string
+	replicaSetName string
+	logger         *zap.Logger
+	timeout        time.Duration
```

Comment:
So much of this is just copied over from the config. Maybe just save the config here instead to simplify things.